nano/node/node.cpp

```diff
@@ -1261,72 +1258,153 @@ nano::signature_checker::~signature_checker ()
-	stopped = true;
-	lock.unlock ();
-	condition.notify_all ();
-	if (thread.joinable ())
+	std::lock_guard<std::mutex> guard (stopped_mutex);
+	if (!stopped)
 	{
-		thread.join ();
+		stopped = true;
+		thread_pool.join ();
 	}
 }
 
 void nano::signature_checker::flush ()
 {
-	std::unique_lock<std::mutex> lock (mutex);
-	while (!stopped && !checks.empty ())
-	{
-		condition.wait (lock);
-	}
+	std::lock_guard<std::mutex> guard (stopped_mutex);
+	while (!stopped && tasks_remaining != 0)
+		;
 }
 
-void nano::signature_checker::verify (nano::signature_check_set & check_a)
+bool nano::signature_checker::verify_batch (const nano::signature_check_set & check_a, size_t start_index, size_t size)
 {
-	/* Verifications is vector if signatures check results
-	 validate_message_batch returing "true" if there are at least 1 invalid signature */
-	auto code (nano::validate_message_batch (check_a.messages, check_a.message_lengths, check_a.pub_keys, check_a.signatures, check_a.size, check_a.verifications));
+	/* Returns false if there are at least 1 invalid signature */
+	auto code (nano::validate_message_batch (check_a.messages + start_index, check_a.message_lengths + start_index, check_a.pub_keys + start_index, check_a.signatures + start_index, size, check_a.verifications + start_index));
 	(void)code;
-	release_assert (std::all_of (check_a.verifications, check_a.verifications + check_a.size, [](int verification) { return verification == 0 || verification == 1; }));
-	check_a.promise->set_value ();
+
+	return std::all_of (check_a.verifications + start_index, check_a.verifications + start_index + size, [](int verification) { return verification == 0 || verification == 1; });
 }
 
-void nano::signature_checker::run ()
+/* This operates on a number of signatures of size (num_batches * batch_size) from the beginning of the check_a pointers.
+ * Caller should check the value of the promise which indicateswhen the work has been completed.
+ */
+void nano::signature_checker::verify_async (nano::signature_check_set & check_a, size_t num_batches, std::promise<void> & promise)
 {
-	nano::thread_role::set (nano::thread_role::name::signature_checking);
-	std::unique_lock<std::mutex> lock (mutex);
-	started = true;
+	auto task = std::make_shared<Task> (check_a, num_batches);
+	++tasks_remaining;
 
-	lock.unlock ();
-	condition.notify_all ();
-	lock.lock ();
+	for (size_t batch = 0; batch < num_batches; ++batch)
+	{
+		auto size = batch_size;
+		auto start_index = batch * batch_size;
 
-	while (!stopped)
+		boost::asio::post (thread_pool, [this, task, size, start_index, &promise] {
+			auto result = this->verify_batch (task->check, start_index, size);
+			release_assert (result);
+
+			if (--task->pending == 0)
+			{
+				--tasks_remaining;
+				promise.set_value ();
+			}
+		});
+	}
+}
+
+// Set the names of all the threads in the thread pool for easier identification
+void nano::signature_checker::set_thread_names (unsigned num_threads)
```

Comment:
It might be simpler to just call:

```
nano::thread_role::set (nano::thread_role::name::signature_checking);
```

within the lambda in `verify_async()` (above) and update `thread_role::set()` to be idempotent

Reply:
yeh that's possible but it might pollute/distract readers from the hot path code in verify_async